R/mlr_graphs_elements.R

```diff
@@ -0,0 +1,118 @@
+#' @include mlr_graphs.R
+
+#' @title Robustify a learner
+#'
+#' @description
+#' Creates a [`Graph`] that can be used to robustify any subsequent learner.
+#' Performs the following steps:
+#' * Drops empty factor levels using [`PipeOpFixFactors`]
+#' * Imputes `numeric` features using [`PipeOpImputeHist`] and [`PipeOpMissInd`]
+#' * Imputes `factor` features using [`PipeOpImputeNewlvl`]
+#' * Encodes `factors` using `one-hot-encoding`. Factors with a cardinality > max_cardinality` are
+#'   collapsed using [`PipeOpCollapseFactors`].
+#' * If `scaling`, numeric features are scaled to mean 0 and standard deviation 1.
+#'
+#' The graph is built conservatively, i.e. the function always tries to assure everything works.
+#' If a learner is provided, some steps can be left out, i.e. if the learner can deal with
+#' factor variables, no encoding is performed.
+#'
+#' @param task [`Task`] \cr
+#'   A task to create a robustifying pipeline for.
+#' @param learner [`Learner`] \cr
+#'   A learner to create a robustifying pipeline for. Optional, if omitted,
+#'   a more conservative pipeline is built.
+#' @param scaling [`logical`] \cr
+#'   Should the data be scaled? Defaults to true.
+#' @param max_cardinality [`integer`] \cr
+#'   Maximum number of factor levels allowed. See above. Default: 1000.
+#' @return Returns a [`Graph`]
+#' @export
+#' @examples
+#' library(mlr3)
+#' lrn = lrn("regr.rpart")
+#' task = mlr_tasks$get("boston_housing")
+#' gr = robustify_pipeline(task, lrn) %>>% po("learner", lrn)
+#' resample(task,GraphLearner$new(gr), rsmp("holdout"))
+robustify_pipeline = function(task, learner = NULL, scaling = TRUE, max_cardinality = 1000) {
```

Comment:
maybe make task optional and assume worst case if not given?

Reply:
also a suggestion: have the arguments
```r
function(task = NULL, learner = NULL, impute_missings = NULL, factors_to_numeric = NULL, ....)
```
and impute missings if `impute_missings` is `TRUE`, of if (`impute_missings` is `NULL` and (task is `NULL` or has `missings`) and (learner is `NULL` or can not handle missings)); similarly with factor to numeric

Reply:
Why do ` impute_missings = NULL, factors_to_numeric = NULL` default to NULL?

Reply:
The idea is that `impute_missings` and `factors_to_numerics` may override the `task` / `learner` properties.

Reply:
You could also do something like
```r
function(task = NULL, learner = NULL, 
  impute_missings = (is.null(task) || any(task$missings())) && (is.null(learner) || !"missings" %in% learner$properties),
  factors_to_numeric = (is.null(task) || <TASK_HAS_FACTORS>) && (is.null(learner) || !"factors" %in% learner$properties))
```